fix(feg): fix snowflake issue in feg integ test

wolfseb opened the pull request:
## Summary

fixes #13540 

Merge after #13781 

This PR creates a file `run.py`, which is to be used instead of `docker-compose up -d` to start the FeG containers. It creates the snowflake file in `/magma/.cache/feg/` if it does not exist already (if there is a directory `snowflake` in its place instead, it removes it first). Further related changes in this PR:
- adapt the automatic and semi-automatic build processes to use `./run.py`
- change the documentation of the manual build process accordingly
- remove a superfluous creation of a snowflake file in the wrong directory in the ci workflow
- remove an unused, empty snowflake file in `lte/gateway/python/.cache/feg/`

## Test Plan

- Spin up the AGW, FeG, and orc8r and run the FeG integ tests and the connectivity test (automatic, semiautomatic, and manual should all work).
- Check that if there is a folder `magma/.cache/feg/snowflake/`, the script will now automatically replace it with a snowflake file, which gets filled with an UUID.

## Additional Information

- [ ] This change is backwards-breaking

maxhbr commented on lte/gateway/python/integ_tests/federated_tests/docker/run.py:
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
```
not sure how this python script compares against the following lines of bash. 

```bash
#!/usr/bin/env bash

SNOWFLAKE_FILE='$MAGMA_ROOT/.cache/feg/snowflake'
SNOWFLAKE_DIR="$(dirname "$SNOWFLAKE_FILE")"
rm -rf "$SNOWFLAKE_DIR"
mkdir -p "$SNOWFLAKE_DIR"
touch "$SNOWFLAKE_FILE"

if [[ $# -eq 0 ]] ; then
  docker-compose up -d
else
  docker-compose "$@"
fi
```

wolfseb replied:
The only logical difference would be that the bash script deletes the `.cache/feg` directory by default and then creates the path and snowflake file, whereas the python script only does anything regarding the snowflake file if it doesn't exist in the correct way yet. But since it is just a cache it probably doesn't matter in the end, unless there is something else in the `feg` folder there (which, to my knowledge, there is not so far).
I made a python script to keep it consistent with how the containers are started for the orc8r (it just seemed convenient to have the same commands to do the same things for different parts of the project).

wolfseb replied:
How about a slightly adapted version:
```bash
#!/usr/bin/env bash

SNOWFLAKE_DIR="$MAGMA_ROOT/.cache/feg"
SNOWFLAKE_FILE="$SNOWFLAKE_DIR/snowflake"

if [[ ! -f "$SNOWFLAKE_FILE" ]] ; then
  if [[ -d "$SNOWFLAKE_FILE" ]] ; then
    rm -rf "$SNOWFLAKE_FILE"
  fi
  mkdir -p "$SNOWFLAKE_DIR"
  touch "$SNOWFLAKE_FILE"
fi

if [[ $# -eq 0 ]] ; then
  docker-compose up -d
else
  docker-compose "$@"
fi
```

sebathomas replied:
Both Python or Bash would be fine with me. In this case, the Bash version seems a little easier. On the other hand, we have a `run.py` file for other components too.

wolfseb replied:
After discussion we decided to stick with the python script, to stay consistent with different parts of the project.

Commit:
chore(feg): remove superfluous snowflake file

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
chore(ci): remove superfluous creation of a snowflake file

Removes the snowflake file creation in $MAGMA_ROOT during the ci integ
test. The correct snowflake file is still being created.

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
fix(feg): fix snowflake file issue in feg integ test

Fixes the incorrect creation of the snowflake file by introducing a
`run.py` executable instead of using `docker-compose up -d`, which
creates the file before docker tries to mount it. The usage is also
changed in the corresponding fabfile for the automated workflow.

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

voisey commented on lte/gateway/python/integ_tests/federated_tests/docker/run.py:
```diff
@@ -0,0 +1,69 @@
+        _exec_cmd(['rm', '-r', SNOWFLAKE_FILE])
+    print("Creating snowflake file")
+    _exec_cmd(['mkdir', '-p', SNOWFLAKE_PATH])
+    _exec_cmd(['touch', SNOWFLAKE_FILE])
```
At what point does the file get filled with a UUID?

wolfseb replied:
To my understanding that is done by the magmad services on the feg via the snowflake.py package, i.e. not anywhere in the touched code. 
You can check that by e.g. replacing the snowflake file with a directory and checking the logs of the magmad container on the feg. It will throw an error:
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/dist-packages/magma/magmad/main.py", line 316, in <module>
    main()
  File "/usr/local/lib/python3.8/dist-packages/magma/magmad/main.py", line 61, in main
    logging.info('Starting magmad for UUID: %s', snowflake.make_snowflake())
  File "/usr/local/lib/python3.8/dist-packages/snowflake.py", line 43, in make_snowflake
    _write_new_id(snowflake_file)
  File "/usr/local/lib/python3.8/dist-packages/snowflake.py", line 33, in _write_new_id
    f = open(snowflake_file, "w")
IsADirectoryError: [Errno 21] Is a directory: '/etc/snowflake'
```

sebathomas replied:
Yes, it's filled by the Python library if there's no folder there.

voisey replied:
Great, thanks!

ajahl commented on lte/gateway/python/integ_tests/federated_tests/docker/run.py:
```diff
@@ -0,0 +1,69 @@
+import subprocess
+
+SNOWFLAKE_PATH = '../../../../../../.cache/feg/'
+SNOWFLAKE_FILE = '../../../../../../.cache/feg/snowflake'
```
Can we replace  `../../../../../..` with MAGMA_ROOT?
Something like `os.environ["MAGMA_ROOT"]`

wolfseb replied:
Good idea, I changed the code accordingly.

Commit:
docs: add changed commands to feg integ test documentation

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

ulaskozat reviewed (approved):
LGTM
